File: app/src/main/java/com/erlangshen/mvp/presenter/NewsListPresenter.java
package com.erlangshen.mvp.presenter;

import com.erlangshen.base.BasePresenter;
import com.erlangshen.base.Constant;
import com.erlangshen.mvp.model.LatestData;
import com.erlangshen.mvp.view.INewListView;
import com.erlangshen.retrofit.ApiStores;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * 新闻列表
 */
public class NewsListPresenter extends BasePresenter<INewListView>{
    public NewsListPresenter(INewListView iView) {
        attachView(iView);
    }

    public void requestNewsListData() {
        getView().showLoading();
        Observable<LatestData> latestData = apiStores.getLatestData("latest");
        addSubscription(latestData,new DisposableObserver<LatestData>() {
            @Override
            public void onNext(LatestData latestData) {
                getView().loadNewsList(latestData.getStories());
            }

            @Override
            public void onError(Throwable e) {
                getView().onError(e);
                getView().hideLoading();
            }

            @Override
            public void onComplete() {
                getView().onSuccess("请求成功!");
                getView().hideLoading();
            }
        });
    }
}

File: app/src/main/java/com/erlangshen/mvp/presenter/NewsListPresenter.kt
package com.erlangshen.mvp.presenter

import com.erlangshen.base.BasePresenter
import com.erlangshen.mvp.model.LatestData
import com.erlangshen.mvp.view.INewsListView
import com.erlangshen.utils.DateUtils
import io.reactivex.Observable
import io.reactivex.observers.DisposableObserver
import javax.inject.Inject

/**
 * 新闻列表
 */
class NewsListPresenter(iView: INewsListView) : BasePresenter<INewsListView>() {
    init {
        attachView(iView)
    }

    var dataList: MutableList<LatestData.StoriesEntity> = mutableListOf()
    @Inject
    lateinit var bihuData: LatestData
    var date: String = ""
    var isRequestDataOk = true
    /**
     * 请求最新新闻
     */
    fun requestNewsListData() {
        getView().showLoading()
        val latestData = apiStores.getLatestData("latest")
        addSubscription(latestData as Observable<*>, object : DisposableObserver<LatestData>() {
            override fun onNext(latestData: LatestData) {
                bihuData = latestData
                dataList.clear()
                dataList.addAll(latestData.stories!!)
                getView().loadNewsList(bihuData)
            }

            override fun onError(e: Throwable) {
                getView().onError(e)
                getView().hideLoading()
            }

            override fun onComplete() {
                getView().onSuccess("请求成功!")
                getView().hideLoading()
            }
        })
    }

    /**
     * 请求往日新闻
     */
    fun requestBeforeData() {
        getView().showLoading()
        if (isRequestDataOk) {
            date = DateUtils.getBeforeDay(date)
            var entity = LatestData.StoriesEntity()
            entity.title = date
            entity.isDate = true
            dataList.add(entity)
        }
        val latestData = apiStores.getBeforeData(date)
        addSubscription(latestData as Observable<*>, object : DisposableObserver<LatestData>() {
            override fun onNext(latestData: LatestData) {
                dataList.addAll(latestData.stories!!)
                bihuData.stories = dataList
                getView().loadBeforeData(bihuData)
            }

            override fun onError(e: Throwable) {
                getView().onError(e)
                getView().hideLoading()
                isRequestDataOk = false
            }

            override fun onComplete() {
                getView().onSuccess("请求成功!")
                getView().hideLoading()
                isRequestDataOk = true
            }
        })
    }
}
